seqr/views/utils/terra_api_utils.py

```diff
@@ -0,0 +1,220 @@
+# This module provides python bindings for the AnVIL Terra API.
+
+import json
+import logging
+import time
+import requests
+
+from urllib.parse import urljoin
+
+from social_django.models import UserSocialAuth
+from social_django.utils import load_strategy
+
+from google.auth.transport.requests import AuthorizedSession
+from google.auth.transport import DEFAULT_REFRESH_STATUS_CODES
+from google.oauth2 import service_account
+
+from settings import SEQR_VERSION, TERRA_API_ROOT_URL, GOOGLE_SERVICE_ACCOUNT_INFO, SOCIAL_AUTH_GOOGLE_OAUTH2_SCOPE
+
+SEQR_USER_AGENT = "seqr/" + SEQR_VERSION
+
+logger = logging.getLogger(__name__)
+
+
+def anvil_enabled():
+    return bool(TERRA_API_ROOT_URL)
+
+
+class TerraAPIException(Exception):
+
+    """For exceptions happen in Terra API calls."""
+
+    pass
+
+
+def _get_call_args(path, headers=None, root_url=None):
+    if headers is None:
+        headers = {"User-Agent": SEQR_USER_AGENT}
+    headers.update({"accept": "application/json"})
+    if root_url is None:
+        root_url = TERRA_API_ROOT_URL
+    url = urljoin(root_url, path)
+    return url, headers
+
+
+class ServiceAccountSession(AuthorizedSession):
+
+    def __init__(self):
+        """Init the session start time to None."""
+        self.started_at = None
+
+    def create_session(self):
+        """Create a service account session."""
+        credentials = service_account.Credentials.from_service_account_info(GOOGLE_SERVICE_ACCOUNT_INFO,
+                                        scopes = SOCIAL_AUTH_GOOGLE_OAUTH2_SCOPE)
+        super(ServiceAccountSession, self).__init__(credentials)
+        self.started_at = time.time()
+
+    def _make_request(self, method, path, headers, root_url, **kwargs):
+        if not anvil_enabled():
+            raise TerraAPIException('AnVIL access is not enabled')
+
+        if not GOOGLE_SERVICE_ACCOUNT_INFO:
+            return {}
```

Comment:
This only happens in testing cases of having `TERRA_API_ROOT_URL` but no mocked responses.

Reply:
no, tests should fail if that happens because `TERRA_API_ROOT_URL` should only be configured for the test when we are trying to test anvil behavior